zebra-state/src/service/read/difficulty.rs

```diff
@@ -38,6 +38,60 @@ pub fn difficulty_and_time_info(
     difficulty_and_time(relevant_chain, tip, network)
 }
 
+/// The default window size specifying how many blocks to check when estimating the chain's solution rate
+/// Based on default value in zcashd.
+const DEFAULT_SOLUTION_RATE_WINDOW_SIZE: usize = 120;
+
+/// Accepts a `non_finalized_state`, [`ZebraDb`], `num_blocks`, and a block hash to start at.
+///
+/// Iterates over up to the last `num_blocks` or [`DEFAULT_SOLUTION_RATE_WINDOW_SIZE`] blocks, summing up
+/// their total work. Divides that total by the number of seconds between the timestamp of the
+/// first block in the iteration and 1 block below the last block.
+///
+/// Returns the solution rate per second for the current best chain, or `None` if the `start_hash` and
+/// at least 1 block below it are not found in the chain.
+pub fn solution_rate(
+    non_finalized_state: &NonFinalizedState,
+    db: &ZebraDb,
+    num_blocks: Option<usize>,
+    start_hash: Hash,
+) -> Option<u128> {
+    let mut total_work: u128 = 0;
+    let num_blocks = num_blocks.unwrap_or(DEFAULT_SOLUTION_RATE_WINDOW_SIZE);
+
+    let mut block_iter = any_ancestor_blocks(non_finalized_state, db, start_hash)
+        .take(num_blocks.checked_add(1).unwrap_or(num_blocks))
```

Comment:
Do we need this extra block? Please document where this comes from.

Reply:
> Do we need this extra block?

It's needed to get the number of seconds between when mining on the first block in the `num_blocks` range likely started. Added a comment explaining in [`333ef9e`](https://github.com/ZcashFoundation/zebra/pull/5808/commits/333ef9ee05e68ccf9b70d331d339e624635c8ed7).